shielder/cli/src/ink_contract.rs

```diff
@@ -0,0 +1,291 @@
+    InkEnv(String),
+    TokenIdAlreadyRegistered(),
+    TokenIdNotRegistered(),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+pub enum OwnableError {
+    CallerIsNotOwner(),
+    NewOwnerIsZero(),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+pub enum BabyLiminalError {
+    IdentifierAlreadyInUse(),
+    VerificationKeyTooLong(),
+    StoreKeyErrorUnknown(),
+    UnknownVerificationKeyIdentifier(),
+    DeserializingProofFailed(),
+    DeserializingPublicInputFailed(),
+    DeserializingVerificationKeyFailed(),
+    VerificationFailed(),
+    IncorrectProof(),
+    VerifyErrorUnknown(),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+pub enum PSP22Error {
+    Custom(Vec<u8>),
+    InsufficientBalance(),
+    InsufficientAllowance(),
+    ZeroRecipientAddress(),
+    ZeroSenderAddress(),
+    SafeTransferCheckFailed(Vec<u8>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+pub enum Relation {
+    Deposit(),
+    DepositAndMerge(),
+    Merge(),
+    Withdraw(),
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct Instance {
+    account_id: ink_primitives::AccountId,
+}
+
+impl From<ink_primitives::AccountId> for Instance {
+    fn from(account_id: ink_primitives::AccountId) -> Self {
+        Self { account_id }
+    }
+}
+
+impl From<Instance> for ink_primitives::AccountId {
+    fn from(instance: Instance) -> Self {
+        instance.account_id
+    }
+}
+
+impl Instance {
+    /// Instantiate the contract. Set the caller as the owner.
+    #[allow(dead_code, clippy::too_many_arguments)]
+    pub async fn new<TxInfo, E, C: ink_wrapper_types::SignedConnection<TxInfo, E>>(
+        conn: &C,
+        salt: Vec<u8>,
+        max_leaves: u32,
+    ) -> Result<Self, E> {
+        let data = {
+            let mut data = vec![155, 174, 157, 94];
+            max_leaves.encode_to(&mut data);
+            data
+        };
+        let code_hash = [
+            149, 184, 4, 182, 7, 123, 213, 168, 101, 221, 109, 139, 50, 45, 79, 143, 182, 54, 121,
+            91, 181, 8, 218, 161, 66, 102, 203, 137, 238, 115, 255, 180,
+        ];
+        let account_id = conn.instantiate(code_hash, salt, data).await?;
+        Ok(Self { account_id })
+    }
+
+    ///  Trigger deposit action (see ADR for detailed description).
+    #[allow(dead_code, clippy::too_many_arguments)]
+    pub async fn deposit<TxInfo, E, C: ink_wrapper_types::SignedConnection<TxInfo, E>>(
+        &self,
+        conn: &C,
+        token_id: u16,
+        value: u128,
+        note: [u64; 4],
+        proof: Vec<u8>,
+    ) -> Result<TxInfo, E> {
+        let data = {
+            let mut data = vec![0, 0, 0, 1];
+            token_id.encode_to(&mut data);
+            value.encode_to(&mut data);
+            note.encode_to(&mut data);
+            proof.encode_to(&mut data);
+            data
+        };
+        conn.exec(self.account_id, data).await
```

Comment:
this is the first time I see output from ink-wrapper: @obrok amazing work! totally love it

Reply:
Same here, I'm speechless :astonished:

Reply:
<img src="https://media2.giphy.com/media/v1.Y2lkPTc5MGI3NjExZWY1MGVmYTk2NWM5OTUwZTVlYjFhN2E1NWJhYTAyYjU4ZDRkYTcwZCZjdD1n/3oEhn7OTiXi4mYReyQ/giphy.gif"/>